src/main/java/org/jitsi/srtp/crypto/OpenSslAesCipherSpi.java

```diff
@@ -0,0 +1,726 @@
+                return inputLen + tagLen;
+            }
+            else
+            {
+                return inputLen;
+            }
+        }
+        else
+        {
+            int len = buffered + inputLen - tagLen;
+            if (len < 0)
+            {
+                return 0;
+            }
+            return len;
+        }
+    }
+
+    @Override
+    protected int engineGetOutputSize(int inputLen)
+    {
+        return getOutputSize(inputLen, true);
+    }
+
+    @Override
+    protected byte[] engineGetIV()
+    {
+        return Arrays.copyOf(iv, iv.length);
+    }
+
+    @Override
+    protected AlgorithmParameters engineGetParameters()
+    {
+        /* TODO: do we need this? */
+        return null;
+    }
+
+    @Override
+    protected void engineInit(int opmode, Key key, SecureRandom random)
+        throws InvalidKeyException
+    {
+        try
+        {
+            engineInit(opmode, key, (AlgorithmParameterSpec) null, random);
+        }
+        catch (InvalidAlgorithmParameterException e)
+        {
+            throw new InvalidKeyException("could not create params", e);
+        }
+    }
+
+    @Override
+    protected void engineInit(int opmode, Key key,
+        AlgorithmParameterSpec params, SecureRandom random)
+        throws InvalidKeyException, InvalidAlgorithmParameterException
+    {
+        if (cipherMode == CipherMode.Unspecified)
+        {
+            throw new IllegalStateException("Cipher mode has not been initialized");
+        }
+
+        if (!key.getAlgorithm().equalsIgnoreCase("AES")
+            || !key.getFormat().equalsIgnoreCase("RAW")
+            || (key.getEncoded().length != 16
+            && key.getEncoded().length != 24
+            && key.getEncoded().length != 32))
+        {
+            throw new InvalidKeyException(
+                "AES SecretKeySpec expected, got " + key.getEncoded().length
+                    + " " + key.getAlgorithm() + "/" + key.getFormat());
+        }
+
+        int openSslEncryptMode;
+        switch (opmode)
+        {
+        case Cipher.ENCRYPT_MODE:
+            openSslEncryptMode = 1;
+            break;
+        case Cipher.DECRYPT_MODE:
+            openSslEncryptMode = 0;
+            break;
+        default:
+            throw new InvalidAlgorithmParameterException("Unsupported opmode " + opmode);
+        }
+        this.opmode = opmode;
+
+        int newIvLen = 0;
+
+        if (params != null)
+        {
+            if (cipherMode == CipherMode.GCM)
+            {
+                if (params instanceof GCMParameterSpec)
+                {
+                    if (((GCMParameterSpec)params).getTLen() != 128)
+                    {
+                        /* ?? Do we need to enforce this? */
+                        throw new InvalidAlgorithmParameterException
+                            ("Unsupported GCM tag length: must be 128");
+                    }
+                    tagLen = ((GCMParameterSpec)params).getTLen() / 8;
```

Comment:
Can the `buffer` array be initialized here? This could remove the null check in doGcmDecryptUpdateWithBuffer